### PR TITLE
Allow overwriting max memory on module instantiation

### PR DESCRIPTION
**Pull request by eloparco**
This PR adds a `max_memory_pages` parameter to module instantiation APIs, to allow overriding the max memory defined in the WASM module.

Sticking to the max memory defined in the module is quite limiting when using shared memory in production. If targeted devices have different memory constraints, many wasm files have to be generated with different max memory values. And device constraints may not be known in advance.

Being able to set the max memory value during module instantiation allows to reuse the same wasm module, e.g. by retrying instantiation with different max memory value.

### CODE REVIEW

File: core/iwasm/aot/aot_runtime.c [hunk 1]
```diff
@@ -783,12 +783,14 @@ static AOTMemoryInstance *
...
-                   char *error_buf, uint32 error_buf_size)
+                   uint32 max_memory_pages, char *error_buf,
+                   uint32 error_buf_size)
 {
     void *heap_handle;
     uint32 num_bytes_per_page = memory->num_bytes_per_page;
     uint32 init_page_count = memory->mem_init_page_count;
-    uint32 max_page_count = memory->mem_max_page_count;
+    uint32 max_page_count =
+        max_memory_pages == 0 ? memory->mem_max_page_count : max_memory_pages;
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:793]: max_page_count should be no smaller than init_page_count, how about:
```C
uint32 max_page_count =
        max_memory_pages == 0 ? memory->mem_max_page_count :
        (max_memory_pages > init_page_count ? max_memory_pages : init_page_count);
```

File: core/iwasm/include/wasm_export.h [hunk 2]
```diff
@@ -527,6 +527,32 @@ wasm_runtime_instantiate(const wasm_module_t module,
...
+ * @param max_memory_pages the number of pages (64k each) to use as maximum memory
+ * @param error_buf buffer to output the error info if failed
+ * @param error_buf_size the size of the error buffer
+ *
+ * @return return the instantiated WASM module instance, NULL if failed
+ */
+WASM_RUNTIME_API_EXTERN wasm_module_inst_t
+wasm_runtime_instantiate_ex(const wasm_module_t module,
+                         uint32_t default_stack_size, uint32_t host_managed_heap_size,
+                         uint32_t max_memory_pages,
```
Comment (wenyongh) [inline core/iwasm/include/wasm_export.h:553]: We may add more arguments in the future and had better not change the API prototype again, how about defining an `InstantiationArgs` struct, and here:
```C
wasm_runtime_instantiate_ex(const wasm_module_t module, const InstantiationArgs *args, char *error_buf, uint32_t error_buf_size);
```

File: core/iwasm/include/wasm_export.h [hunk 3]
```diff
@@ -527,6 +527,32 @@ wasm_runtime_instantiate(const wasm_module_t module,
...
+ *        create the operation stack internally with the stack size specified
+ *        here. And API wasm_runtime_create_exec_env() creates the operation
+ *        stack with stack size specified by its parameter, the stack size
+ *        specified here is ignored.
+ * @param host_managed_heap_size the default heap size of the module instance, a heap will
+ *        be created besides the app memory space. Both wasm app and native
+ *        function can allocate memory from the heap.
+ * @param max_memory_pages the number of pages (64k each) to use as maximum memory
+ * @param error_buf buffer to output the error info if failed
+ * @param error_buf_size the size of the error buffer
```
Comment (wenyongh) [inline core/iwasm/include/wasm_export.h:546]: Many duplicated comments with API wasm_runtime_instantiate, how about just mentioning same as it with some argument changes? Refer to
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/include/wasm_export.h#L489

File: core/iwasm/interpreter/wasm_runtime.c [hunk 4]
```diff
@@ -374,7 +375,9 @@ memories_instantiate(const WASMModule *module, WASMModuleInstance *module_inst,
     for (i = 0; i < module->import_memory_count; i++, import++, memory++) {
         uint32 num_bytes_per_page = import->u.memory.num_bytes_per_page;
         uint32 init_page_count = import->u.memory.init_page_count;
-        uint32 max_page_count = import->u.memory.max_page_count;
+        uint32 max_page_count = max_memory_pages == 0
+                                    ? import->u.memory.max_page_count
+                                    : max_memory_pages;
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:380]: max_page_count should be no smaller than init_page_count, had better add a check like comment in aot_runtime.c?

File: core/iwasm/interpreter/wasm_runtime.c [hunk 5]
```diff
@@ -412,12 +415,15 @@ memories_instantiate(const WASMModule *module, WASMModuleInstance *module_inst,
 
     /* instantiate memories from memory section */
     for (i = 0; i < module->memory_count; i++, memory++) {
+        uint32 max_page_count = max_memory_pages == 0
+                                    ? module->memories[i].max_page_count
+                                    : max_memory_pages;
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:420]: Same as above.

**Comment by lum1n0us** [comment]
```
    uint32 max_page_count =
        max_memory_pages == 0 ? memory->mem_max_page_count : max_memory_pages;
```

Does it suggest that the `max_memory_pages` (specified by API) can be greater than `mem_max_page_count` from memory limitation?

**Comment by eloparco** [comment]
> Does it suggest that the max_memory_pages (specified by API) can be greater than mem_max_page_count from memory limitation?

Is that not allowed? Actually I remember that was mentioned https://bytecodealliance.zulipchat.com/#narrow/stream/349267-wasi-threads/topic/rust/near/402263578, but couldn't find it in the specs.

**Comment by wenyongh** [comment]
> > Does it suggest that the max_memory_pages (specified by API) can be greater than mem_max_page_count from memory limitation?
> 
> Is that not allowed? Actually I remember that was mentioned https://bytecodealliance.zulipchat.com/#narrow/stream/349267-wasi-threads/topic/rust/near/402263578, but couldn't find it in the specs.

No, from the comment of Alex, the max_memory_pages specified by the API can be smaller than what it is specified in wasm file, but should not be larger than it. So how about we add checks: if max_memory_pages from API is smaller than memory->init_page_count, then set it memory->init_page_count; and if max_memory_pages is larger than memory->max_page_count, then set it to memory->max_page_count?
@lum1n0us what's your opinion?

**Comment by eloparco** [comment]
> So how about we add checks: if max_memory_pages from API is smaller than memory->init_page_count, then set it memory->init_page_count; and if max_memory_pages is larger than memory->max_page_count, then set it to memory->max_page_count?

Yeah, makes sense, I'll change the code to do that

File: core/iwasm/include/wasm_export.h (hunk 2 above)
Comment (eloparco) [reply core/iwasm/include/wasm_export.h]: Good idea

File: core/iwasm/aot/aot_runtime.c [hunk 6]
```diff
@@ -783,12 +783,18 @@ static AOTMemoryInstance *
...
+                   uint32 max_memory_pages, char *error_buf,
+                   uint32 error_buf_size)
 {
     void *heap_handle;
     uint32 num_bytes_per_page = memory->num_bytes_per_page;
     uint32 init_page_count = memory->mem_init_page_count;
-    uint32 max_page_count = memory->mem_max_page_count;
+    uint32 max_page_count =
+        max_memory_pages == 0
+            ? memory->mem_max_page_count
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:794]: How about handling `max_memory_pages == 0` inside `wasm_runtime_get_overwritten_max_mem`, and here it just calls `wasm_runtime_get_overwritten_max_mem`?

File: core/iwasm/common/wasm_runtime_common.c [hunk 7]
```diff
@@ -1357,24 +1357,44 @@ wasm_runtime_unload(WASMModuleCommon *module)
 #endif
 }
 
+uint32
+wasm_runtime_get_overwritten_max_mem(uint32 max_memory_pages,
+                                     uint32 module_init_page_count,
+                                     uint32 module_max_page_count)
+{
```
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.c:1364]: How about adding
```C
    if (max_memory_pages == 0) {
        return module_max_page_count;
    }
```

File: core/iwasm/common/wasm_runtime_common.c [hunk 8]
```diff
@@ -1357,24 +1357,44 @@ wasm_runtime_unload(WASMModuleCommon *module)
...
+    if (max_memory_pages < module_init_page_count) {
+        LOG_WARNING("Cannot override max memory with value lower than module "
```
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.c:1366]: Not sure is it better to use `LOG_WARNING("warning: ...")`?

File: core/iwasm/interpreter/wasm_runtime.c [hunk 9]
```diff
@@ -374,7 +375,12 @@ memories_instantiate(const WASMModule *module, WASMModuleInstance *module_inst,
     for (i = 0; i < module->import_memory_count; i++, import++, memory++) {
         uint32 num_bytes_per_page = import->u.memory.num_bytes_per_page;
         uint32 init_page_count = import->u.memory.init_page_count;
-        uint32 max_page_count = import->u.memory.max_page_count;
+        uint32 max_page_count =
+            max_memory_pages == 0
+                ? import->u.memory.max_page_count
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:380]: How about handling `max_memory_pages == 0` inside `wasm_runtime_get_overwritten_max_mem`, and here it just calls, `wasm_runtime_get_overwritten_max_mem`?

File: core/iwasm/interpreter/wasm_runtime.c [hunk 10]
```diff
@@ -412,12 +418,18 @@ memories_instantiate(const WASMModule *module, WASMModuleInstance *module_inst,
 
     /* instantiate memories from memory section */
     for (i = 0; i < module->memory_count; i++, memory++) {
+        uint32 max_page_count =
+            max_memory_pages == 0
+                ? module->memories[i].max_page_count
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:423]: Same as above

File: core/iwasm/common/wasm_runtime_common.c (hunk 8 above)
Comment (eloparco) [reply core/iwasm/common/wasm_runtime_common.c:1370]: If the user is overriding the logger (i.e. `WAMR_BH_LOG`), they probably already have something to distinguish between different log levels: different colors (red for error, yellow for warning) or different log suffixes (e.g. `W`, `E` suffixes). So adding the log level to the log message is redundant.

Comment (wenyongh) [reply core/iwasm/common/wasm_runtime_common.c:1370]: OK, got it, thanks.

**Review by wenyongh — COMMENTED** [review]
LGTM

**Pull request merged**